File: simplified-app-shared/src/main/java/org/nypl/simplified/app/catalog/CatalogAcquisitionButtonController.java
package org.nypl.simplified.app.catalog;

import android.app.Activity;
import android.app.FragmentManager;
import android.content.Intent;
import android.view.View;
import android.view.View.OnClickListener;

import com.io7m.jfunctional.OptionType;
import com.io7m.jnull.NullCheck;
import com.io7m.jnull.Nullable;
import com.io7m.junreachable.UnreachableCodeException;

import org.nypl.simplified.app.LoginActivity;
import org.nypl.simplified.app.LoginDialog;
import org.nypl.simplified.app.LoginListenerType;
import org.nypl.simplified.app.R;
import org.nypl.simplified.app.Simplified;
import org.nypl.simplified.books.core.AccountBarcode;
import org.nypl.simplified.books.core.AccountCredentials;
import org.nypl.simplified.books.core.AccountGetCachedCredentialsListenerType;
import org.nypl.simplified.books.core.AccountPIN;
import org.nypl.simplified.books.core.BookID;
import org.nypl.simplified.books.core.BooksType;
import org.nypl.simplified.books.core.FeedEntryOPDS;
import org.nypl.simplified.books.core.LogUtilities;
import org.nypl.simplified.opds.core.OPDSAcquisition;
import org.nypl.simplified.opds.core.OPDSAcquisitionPath;
import org.nypl.simplified.opds.core.OPDSAcquisitionRelation;
import org.slf4j.Logger;

/**
 * A controller for an acquisition button.
 *
 * This is responsible for logging in, if necessary, and then starting the
 * borrow of a given book.
 */

public final class CatalogAcquisitionButtonController
  implements OnClickListener, LoginListenerType
{
  private static final Logger LOG;

  static {
    LOG = LogUtilities.getLog(CatalogAcquisitionButtonController.class);
  }

  private final OPDSAcquisitionPath acquisition;
  private final Activity        activity;
  private final BooksType       books;
  private final FeedEntryOPDS   entry;
  private final BookID          id;

  /**
   * Construct a button controller.
   *
   * @param in_activity The host activity
   * @param in_books    The books database
   * @param in_id       The book ID
   * @param in_acq      The acquisition
   * @param in_entry    The associated feed entry
   */

  public CatalogAcquisitionButtonController(
    final Activity in_activity,
    final BooksType in_books,
    final BookID in_id,
    final OPDSAcquisitionPath in_acq,
    final FeedEntryOPDS in_entry)
  {
    this.activity = NullCheck.notNull(in_activity);
    this.acquisition = NullCheck.notNull(in_acq);
    this.id = NullCheck.notNull(in_id);
    this.books = NullCheck.notNull(in_books);
    this.entry = NullCheck.notNull(in_entry);
  }

  @Override public void onClick(
    final @Nullable View v)
  {
    if (this.books.accountIsLoggedIn()
      && Simplified.getCurrentAccount().needsAuth()
      && this.acquisition.getNext().getRelation() != OPDSAcquisitionRelation.ACQUISITION_OPEN_ACCESS) {
      this.books.accountGetCachedLoginDetails(
        new AccountGetCachedCredentialsListenerType()
        {
          @Override public void onAccountIsNotLoggedIn()
          {
            throw new UnreachableCodeException();
          }

          @Override public void onAccountIsLoggedIn(
            final AccountCredentials creds)
          {
            CatalogAcquisitionButtonController.this.onLoginSuccess(creds);
          }
        });
    } else if (!Simplified.getCurrentAccount().needsAuth()
      || this.acquisition.getNext().getRelation() == OPDSAcquisitionRelation.ACQUISITION_OPEN_ACCESS) {
      this.getBook();
    }
    else {
      this.tryLogin();
    }
  }

  private void tryLogin()
  {
    final boolean clever_enabled =
      this.activity.getResources().getBoolean(R.bool.feature_auth_provider_clever);

    if (clever_enabled) {
      final Intent account = new Intent(this.activity, LoginActivity.class);
      this.activity.startActivityForResult(account, 1);
      this.activity.overridePendingTransition(0, 0);
    } else {
      final AccountBarcode barcode = new AccountBarcode("");
      final AccountPIN pin = new AccountPIN("");

      final LoginDialog df = LoginDialog.newDialog("Login required", barcode, pin);
      df.setLoginListener(this);
      final FragmentManager fm = this.activity.getFragmentManager();
      df.show(fm, "login-dialog");
    }
  }

  @Override public void onLoginAborted()
  {
    // Nothing
  }

  @Override public void onLoginFailure(
    final OptionType<Throwable> error,
    final String message)
  {
    // Nothing
  }

  @Override public void onLoginSuccess(
    final AccountCredentials creds)
  {
    LOG.debug("login succeeded");
    this.getBook();
  }

  private void getBook() {
    LOG.debug("attempting borrow of {} acquisition", this.acquisition.getNext().getRelation());

    switch (this.acquisition.getNext().getRelation()) {
      case ACQUISITION_BORROW:
      case ACQUISITION_GENERIC:
      case ACQUISITION_OPEN_ACCESS: {
        this.books.bookBorrow(this.id, this.entry.getFeedEntry(), Simplified.getCurrentAccount().needsAuth());
        break;
      }
      case ACQUISITION_BUY:
      case ACQUISITION_SAMPLE:
      case ACQUISITION_SUBSCRIBE: {
        throw new UnsupportedOperationException();
      }
    }
  }
}

File: simplified-app-shared/src/main/java/org/nypl/simplified/app/catalog/CatalogAcquisitionButtonController.kt
package org.nypl.simplified.app.catalog

import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.View.OnClickListener
import android.widget.Toast

import com.google.common.util.concurrent.ListeningExecutorService
import com.io7m.jfunctional.None
import com.io7m.jfunctional.Option
import com.io7m.jfunctional.OptionType
import com.io7m.jfunctional.OptionVisitorType
import com.io7m.jfunctional.Some
import com.io7m.jnull.Nullable

import org.nypl.simplified.app.NetworkConnectivityType
import org.nypl.simplified.app.R
import org.nypl.simplified.app.login.LoginCancelledType
import org.nypl.simplified.app.login.LoginDialog
import org.nypl.simplified.app.login.LoginFailedType
import org.nypl.simplified.app.login.LoginSucceededType
import org.nypl.simplified.books.accounts.AccountType
import org.nypl.simplified.books.accounts.AccountsDatabaseNonexistentException
import org.nypl.simplified.books.book_database.BookID
import org.nypl.simplified.books.book_registry.BookRegistryReadableType
import org.nypl.simplified.books.book_registry.BookWithStatus
import org.nypl.simplified.books.controller.BooksControllerType
import org.nypl.simplified.books.controller.ProfilesControllerType
import org.nypl.simplified.books.document_store.DocumentStoreType
import org.nypl.simplified.books.feeds.FeedEntry
import org.nypl.simplified.books.logging.LogUtilities
import org.nypl.simplified.books.profiles.ProfileNoneCurrentException
import org.nypl.simplified.opds.core.OPDSAcquisition
import org.slf4j.LoggerFactory

/**
 * A controller for an acquisition button.
 *
 * This is responsible for logging in, if necessary, and then starting the
 * borrow of a given book.
 */

class CatalogAcquisitionButtonController(
  private val acquisition: OPDSAcquisition,
  private val activity: AppCompatActivity,
  private val books: BooksControllerType,
  private val entry: FeedEntry.FeedEntryOPDS,
  private val id: BookID,
  private val profiles: ProfilesControllerType,
  private val bookRegistry: BookRegistryReadableType,
  private val networkConnectivity: NetworkConnectivityType,
  private val backgroundExecutor: ListeningExecutorService,
  private val documents: DocumentStoreType) : OnClickListener {

  override fun onClick(@Nullable v: View) {

    if (!this.networkConnectivity.isNetworkAvailable) {
      Toast.makeText(
        this.activity.applicationContext,
        this.activity.resources.getString(R.string.catalog_added_for_later_download),
        Toast.LENGTH_LONG)
        .show()
    }

    val account = accountForBook(this.profiles, this.bookRegistry, this.id)

    val authentication_required = account.provider().authentication().isSome && this.acquisition.relation !== OPDSAcquisition.Relation.ACQUISITION_OPEN_ACCESS

    val authentication_provided = account.credentials().isSome
    if (authentication_required && !authentication_provided) {
      this.tryLogin(account)
      return
    }

    this.tryBorrow(account)
  }

  private fun tryBorrow(account: AccountType) {
    LOG.debug("trying borrow of type {}", this.acquisition.type)
    this.books.bookBorrow(account, this.id, this.acquisition, this.entry.feedEntry)
  }

  private fun tryLogin(account: AccountType) {
    LOG.debug("trying login")

    // XXX: Untranslated string!
    val dialog = LoginDialog.newDialog(
      this.profiles,
      this.backgroundExecutor,
      this.documents,
      "Login Required",
      account,
      LoginSucceededType { creds -> this.onLoginSuccess(account) },
      LoginCancelledType { this.onLoginCancelled() },
      LoginFailedType { error, message -> this.onLoginFailure(error, message) })

    dialog.show(this.activity.supportFragmentManager, "login-dialog")
  }

  private fun onLoginFailure(error: OptionType<Exception>, message: String) {
    LogUtilities.errorWithOptionalException(LOG, message, error)
  }

  private fun onLoginCancelled() {
    LOG.debug("login cancelled")
  }

  private fun onLoginSuccess(
    account: AccountType) {

    LOG.debug("login succeeded")
    this.tryBorrow(account)
  }

  companion object {

    private val LOG = LoggerFactory.getLogger(CatalogAcquisitionButtonController::class.java)

    private fun accountForBook(
      profiles: ProfilesControllerType,
      book_registry: BookRegistryReadableType,
      book_id: BookID): AccountType {

      return Option.of<BookWithStatus>(book_registry.books()[book_id]).accept(
        object : OptionVisitorType<BookWithStatus, AccountType> {
          override fun none(none: None<BookWithStatus>): AccountType {
            try {
              return profiles.profileAccountCurrent()
            } catch (e: ProfileNoneCurrentException) {
              throw IllegalStateException(e)
            }

          }

          override fun some(some: Some<BookWithStatus>): AccountType {
            try {
              val profile = profiles.profileCurrent()
              val account_id = some.get().book().account
              return profile.account(account_id)
            } catch (e: ProfileNoneCurrentException) {
              throw IllegalStateException(e)
            } catch (e: AccountsDatabaseNonexistentException) {
              throw IllegalStateException(e)
            }

          }
        })
    }
  }
}
